### ISSUE TITLE
False positive `unnecessary-lambda` (wrong type inference in if-else-expression)

### ISSUE THREAD

**Issue by robertschweizer** [pylint-dev/pylint#7647]
### Bug description

This triggers an `unnecessary-lambda` warning:

```py
# pylint: disable=missing-docstring
from contextlib import ExitStack


def fun(**kwargs):
    print(kwargs)

def run_in_context(callable_):
    with ExitStack():  # pytest.raises and pytest.warns in my case
        callable_()

def main(omit_arg: bool):
    kwargs = {} if omit_arg else {"arg": 2}
    run_in_context(lambda: fun(**kwargs))
```

If I replace `kwargs = {} if omit_arg else {"arg": 2}` with `kwargs = {"arg": 2} if not omit_arg else {}`, no warning is shown. 

I assume that the type inference is not working correctly for the if-else conditional expression.

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:14:19: W0108: Lambda may not be necessary (unnecessary-lambda)

------------------------------------------------------------------
Your code has been rated at 8.89/10 (previous run: 9.09/10, -0.20)
```

### Expected behavior

No `unnecessary-lambda` warning should be raised.

### Pylint version

```shell
pylint 2.15.4
astroid 2.12.11
Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)]
```

### OS / Environment

Windows 10, 64bit

**Comment by clavedeluna** [comment]
I indeed tracked this down to astroid's inference system, the checker should've short circuited [here](https://github.com/PyCQA/pylint/blob/main/pylint/checkers/base/basic_checker.py#L529) but it did not because `call_site.keyword_arguments` is not pupulated.
